### ISSUE TITLE
pbench-uperf is_remote incorrectly detects localhost names or addresses as remote

### ISSUE THREAD

**Issue by ldoktor** [distributed-system-analysis/pbench#2635]
The pbench-uperf's `is_remote` function:

```bash
function is_remote {
    # Returns 0 if the argument is considered "remote", and 1 if not.
    [[ "${1}" != "127.0.0.1" && "${1}" != "localhost" ]]
}
```

only detects localhost, but not `hostname`, `::1` or other local interfaces. This wasn't a problem on RHEL8 as later: `scp $PATH $HOSTNAME:$PATH` preserved the local file, but in RHEL9 it empties the file, resulting in empty `*_cmd_file` and `xml_file`.

**Comment by ldoktor** [comment]
Sample output from execution:

```
13:51:51.610: root           : Sending command: benchmark_bin=/usr/bin/uperf PERL5LIB=/opt/pbench-agent/tool-scripts/postprocess/:/opt/pbench-agent/bench-scripts/postprocess/ pbench-uperf  --instances=64 --max-stddev=15 --message-sizes=16384 --protocols=tcp --runtime=60 --samples=3 --test-types=stream --clients=virtlab722.virt.lab.eng.bos.redhat.com --servers virtlab722.virt.lab.eng.bos.redhat.com
13:51:51.952: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:51:52.565: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:52:22.551: virtlab722     : >> Starting iteration  (1 of 1)
13:52:23.054: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:52:23.314: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:52:23.599: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:52:23.855: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:52:24.110: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:52:24.396: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:52:24.671: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:52:25.211: virtlab722     : >> test sample 1 of 3
13:52:25.231: virtlab722     : >> client[virtlab722.virt.lab.eng.bos.redhat.com]protocol[tcp]test[stream]instances[64]size[16384] <-> server[virtlab722.virt.lab.eng.bos.redhat.com]
13:52:25.285: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:53:32.405: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:53:32.680: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:53:32.884: virtlab722     : >> File "/var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample1/client::virtlab722.virt.lab.eng.bos.redhat.com-server::virtlab722.virt.lab.eng.bos.redhat.com:20010--server.log" not found.
13:53:32.885: virtlab722     : >> Failed to download '/var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample1/client::virtlab722.virt.lab.eng.bos.redhat.com-server::virtlab722.virt.lab.eng.bos.redhat.com:20010--server.log'
13:53:32.949: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:53:33.151: virtlab722     : >> File "/var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample1/client::virtlab722.virt.lab.eng.bos.redhat.com-server::virtlab722.virt.lab.eng.bos.redhat.com:20010--client_output.txt" not found.
13:53:33.151: virtlab722     : >> Failed to download '/var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample1/client::virtlab722.virt.lab.eng.bos.redhat.com-server::virtlab722.virt.lab.eng.bos.redhat.com:20010--client_output.txt'
13:53:34.212: virtlab722     : >> uperf-postprocess: could not find any result files to process, exiting
...
```
it runs 3 iterations, then it attempts to restart the iteration and hangs for infinity:

```
13:54:47.592: virtlab722     : >> test sample 3 of 3
13:54:47.612: virtlab722     : >> client[virtlab722.virt.lab.eng.bos.redhat.com]protocol[tcp]test[stream]instances[64]size[16384] <-> server[virtlab722.virt.lab.eng.bos.redhat.com]
13:54:47.664: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:54.756: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:55.032: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:55.237: virtlab722     : >> File "/var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample3/client::virtlab722.virt.lab.eng.bos.redhat.com-server::virtlab722.virt.lab.eng.bos.redhat.com:20010--server.log" not found.
13:55:55.237: virtlab722     : >> Failed to download '/var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample3/client::virtlab722.virt.lab.eng.bos.redhat.com-server::virtlab722.virt.lab.eng.bos.redhat.com:20010--server.log'
13:55:55.301: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:55.502: virtlab722     : >> File "/var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample3/client::virtlab722.virt.lab.eng.bos.redhat.com-server::virtlab722.virt.lab.eng.bos.redhat.com:20010--client_output.txt" not found.
13:55:55.502: virtlab722     : >> Failed to download '/var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample3/client::virtlab722.virt.lab.eng.bos.redhat.com-server::virtlab722.virt.lab.eng.bos.redhat.com:20010--client_output.txt'
13:55:56.578: virtlab722     : >> uperf-postprocess: could not find any result files to process, exiting
13:55:56.620: virtlab722     : >> Can't open /var/lib/pbench-agent/uperf__2022.02.21T13.58.18/1-tcp_stream-16384B-64i/sample1/result.json: No such file or directory at /opt/pbench-agent/bench-scripts/postprocess/process-iteration-samples line 166.
13:55:56.625: virtlab722     : >> Restarting iteration 1-tcp_stream-16384B-64i (1 of 1)
13:55:56.706: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:56.961: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:57.265: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:57.565: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:57.819: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:58.069: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:58.336: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
13:55:58.844: virtlab722     : >> pbench-tool-meister-client: Status message not successful: 'failure communicating with TMs'
13:55:58.863: virtlab722     : >> test sample 1 of 3
13:55:58.884: virtlab722     : >> client[virtlab722.virt.lab.eng.bos.redhat.com]protocol[tcp]test[stream]instances[64]size[16384] <-> server[virtlab722.virt.lab.eng.bos.redhat.com]
13:55:58.936: virtlab722     : >> Warning: Permanently added 'virtlab722.virt.lab.eng.bos.redhat.com' (ED25519) to the list of known hosts.
Aborted by [Lukas Doktor](https://kvmqe-jenkins-csb-rhelvirt.apps.ocp4.prod.psi.redhat.com/user/ldoktor)
Sending interrupt signal to process
07:28:33.518: virtlab722     : >> Shared connection to virtlab722.virt.lab.eng.bos.redhat.com closed.
```

**Comment by ldoktor** [comment]
I verified the issue is the `scp` as it detects virtlab722 as remote location.

**Comment by ldoktor** [comment]
Note concurrently I filed a BZ to find out whether the new behaviour is expected: https://bugzilla.redhat.com/show_bug.cgi?id=2056884

**Comment by dbutenhof** [comment]
Take a look at #2542, which fixes `pbench-fio` and `pbench-uperf` to use an improved localhost recognition mechanism (using `hostname --all-ip-addresses` and Python `ipaddress`) which should be able to recognize any "localhost" alias known to the system (IPv4 or IPv6).

It's been in circulation since last fall, but due to many other issues never got completed; it's been recently dusted off and ought to be in soon.

**Comment by ldoktor** [comment]
I remember seeing the `hostname --all-ip-addresses` somewhere, failed to remember where. I hope your fix gets in soon. In the mean-time there is a possibility to use `scp -O` which should force RHEL9 to work similarly to RHEL8 (reply from the RHBZ)

**Comment by dbutenhof** [comment]
> I remember seeing the `hostname --all-ip-addresses` somewhere, failed to remember where. I hope your fix gets in soon.

A PR on this change just went into `main`; basically, `hostname --ip-address` (which we were using before) says to use `--all-ip-addresses` instead, which doesn't rely on name resolution and omits the link-local IPv6 address (which breaks the Python `ipaddress.ip_address()` parser pre-3.9, as well as providing a very restricted and complicated subnet routing model for our purposes).

> In the mean-time there is a possibility to use `scp -O` which should force RHEL9 to work similarly to RHEL8 (reply from the RHBZ)

In theory you should be able to do this yourself by modifying `/opt/pbench-agent/config/pbench-agent.cfg` (or `pbench-agent-default.cfg` to set the `scp_opts` config variable. The default is `-o BatchMode=yes -o StrictHostKeyChecking=no`. The following should do what you want:

```
scp_opts = -O -o BatchMode=yes -o StrictHostKeyChecking=no
```

**Comment by ldoktor** [comment]
Thank you, the 'scp_opts' way is what I did yesterday and it seems to be working well for now, anyway I'm looking forward to the proper fixed `is_remote` version.

**Comment by ldoktor** [comment]
Just a note to others following this advice, RHEL8, older RHEL9s and Fedora 34 scp does not support the `-O` argument. Anyway on later RHEL9s this fix works well...
